homeassistant/components/sensor/__init__.py

```diff
@@ -358,6 +368,15 @@ def unit_of_measurement(self) -> str | None:
 
         native_unit_of_measurement = self.native_unit_of_measurement
 
+        entity_registry = er.async_get(self.hass)
+        entry = entity_registry.async_get(self.entity_id)
+        if (
+            entry
+            and (sensor_options := entry.options.get(DOMAIN))
+            and (custom_unit := sensor_options.get("unit"))
```

Comment:
would it make sense to fetch this once when `self.registry_entry` is set ?

Reply:
It's set by  `EntityPlatform` before adding the entity and by `Entity` which listens to updates.
We could memoize the value here based on the id of the registry entry maybe?